### DISCUSSION TITLE
Feedback needed on new UI - Bulk actions on prims in the Outliner/Viewport

### DISCUSSION THREAD

**Discussion by womanyoyoyo-adsk** [Autodesk/maya-usd#2301] [Show and tell]
Hi everybody,

I am currently working on a design to help artists make bulk actions on prims. Currently prims can be bulk edited via the Channel Box but the Channel Box is only displaying attributes and not metadata. There are other global USD actions that are not attributes or metadata such as unloading, switching a variant, adding a new prim, or editing as Maya data that we'd like to offer for bulk editing. The Outliner is one place to offer these capabilities. Currently the Outliner menu performs actions on the prim that is right-clicked on to trigger the menu, no matter if there's a selection or not. Like this:

![stools-rmb](https://user-images.githubusercontent.com/49446314/164091616-6e62b648-d54c-4f26-870a-ad88c8563b3a.gif)
_Even though 4 stools are selected, the action 'Deactivate Prim' gets executed on 'stool2', the prim that was right-clicked on._ 

In the example above, maybe artists would expect the action 'Deactivate Prim' to be executed on all selected prims. This is a common design pattern in modern applications. However, the Outliner menu triggering on prims that are right-clicked (no matter if selected or not) can be a good thing for this reason: a user is able to have a selection, right-click on a prim not in the selection, perform an action (like unload or a variant switch), and the selection would still be preserved. Like this:

![fridge-variant](https://user-images.githubusercontent.com/49446314/164093573-0bc7ebb7-8e37-4467-a01d-d6a5d7d47c8c.gif)
_4 stools are selected but the user can RMB on the fridge and switch variants, preserving the selection._

The examples above are what's happening currently in Maya 2023 with MayaUSD 0.16. 

**The Design**

When a user right-clicks on a selected prim in the Outliner, the menu would show a new item at the top called 'Selected Prims' with all available bulk actions that would apply to all selected prims (even across stages). The available bulk actions would be provided by the prim which was right-clicked on. This is illustrated in the the example below - the fridge has two variants that can be switched in bulk for all selected prims, but only the prims with the variant 'PixarFridge' would be affected:

![fridge-bulk](https://user-images.githubusercontent.com/49446314/164105256-30281e97-3d2e-40bd-9286-47b8fbde48a6.gif)
_New 'Selected Prims' menu item at the top with all available bulk actions. Bulk actions would be provided by the prim that was right-clicked on but would apply to all selected prims._

If the user has selected prims but wants to perform an action to only one of the selected prims, he can like this:

![fridge-single](https://user-images.githubusercontent.com/49446314/164255651-d811fef6-0b76-4f0d-b8bd-a0f6d851dbd1.gif)
_User is able to operate on a single prim within a selection, while still preserving the selection._

User can have a selection and perform an action to a deselected prim while still preserving the selection, like this:
 
![mixer](https://user-images.githubusercontent.com/49446314/164256875-0ef5a0d1-204b-429c-8c61-222afeea17fc.gif)
_'mixer' prim is deselected but user can deactivate it while still preserving the selection._

When there is a selection in the viewport and the user picks on a selected item, he can execute an action on the prim he picked or on all the selected prims by going to the new menu item 'Selected Prims'. Bulk actions offered in this submenu would be provided by the prim picked (like the Outliner) like this:

![viewport-stools](https://user-images.githubusercontent.com/49446314/164258747-0bdfd6a6-b915-4594-a200-947cbb84d1b3.gif)
_Viewport menu has new 'Selected Prims' item offering bulk actions for all selected prims._

When there are selected prims in the viewport and the user picks on a deselected prim, he can still execute an action on the prim he picked while preserving his selection like this:

![jar-selected](https://user-images.githubusercontent.com/49446314/164263622-beea0508-a709-49c9-9036-a4587d996c9f.gif)
_Viewport menu for deselected prims. Running an action on a deselected prim would preserve your selection._ 

**Questions**

1. What do you think of the design proposed regarding the Outliner menu?
2. What do you think of the design proposed regarding the Viewport menu?
4. Should prim metadata appear in the Channel Box?
5. When I post designs like this, would you prefer paragraphs and gifs (like this post) or a video with me talking about it?
6. Do you have any other thoughts, feedback or suggestions?

Thanks,
Natalia (designer on MayaUSD)

**Comment by dj-mcg** [comment]
Hey @womanyoyoyo-adsk 

Thanks for writing this up! I think this format works really well for me, but if it's easier for you to post a video I think that would work too.

If the goal is to preserve the existing Maya context menu behavior while introducing bulk actions, I think this is a good approach. I'm assuming that the submenus would fly out on hover, as requiring 4-5 clicks to reach a variant selection would cause our users to complain about RSI risk. I do worry about the level of submenus we're introducing here, as this has already come up as a complaint from our users, but I don't really have a solution to this offhand.

I'll ask @gracekumagai to weigh in with any additional thoughts!

**Reply by womanyoyoyo-adsk** [reply]
The switching variants in bulk is a major concern in this design. Thanks for bringing this up. For now this design is on hold. Keep you posted when I have something new. Thanks @dj-mcg .

**Comment by daviddeacon** [comment]
> What do you think of the design proposed regarding the Outliner menu?

 I'm fine with the proposed sub menu for multi selected objects, it does break convention but I can see it being useful. It will just need some getting used to.

> What do you think of the design proposed regarding the Viewport menu?

The hot box menu, it looks good but it REALLY needs to highlight the prim it's going to be acting on when you activate it. It looks WAY too easy for someone to go in and deactivate half of the environment and not realise how they did it. As their mouse cursor was a pixel off and hit the wrong prim. Also does it respect the USD selection mode?

Some kind of edits list would be really useful or some kind of highlighting in the outliner to any edited prims that haven't been saved.

> Should prim metadata appear in the Channel Box?

Yeah I don't see why not, more data the better.

> When I post designs like this, would you prefer paragraphs and gifs (like this post) or a video with me talking about it?

Formats fine, more in-depth video demos with you talking over your actions are always welcome.

> Do you have any other thoughts, feedback or suggestions?

One thing that bugs the hell out of me is having to have the show shapes in the outliner to see the USD stage in the outliner. It shouldn't need this option at all, or at least need it's own 'Expand USD stage' option. You also need to expand two nodes to see the first prims in the stage, it should only be one...

Thanks for involving the community in this way.

**Reply by womanyoyoyo-adsk** [reply]
Hi @daviddeacon,

Thanks very much for this feedback. I especially appreciate the highlighting the prim when the viewport menu is activated. The future design of this will respect the USD selection mode. 

> Some kind of edits list would be really useful or some kind of highlighting in the outliner to any edited prims that haven't been saved.

I'd like to know more about this. It is not a Maya thing but I see where you're going with this. Can you tell me why an artist would find this useful?

We've scheduled the Show Shapes issue #1454.  

Regarding the above design, I am going to rework it and post an update in the future. It has some issues. I'll keep you posted.

**Comment by dj-mcg** [comment]
Oh! Sorry - I missed one question. My vote for including metadata in the Channel Box would be NO, but that might be specific to our user's workflows. I think the Channel Box could get way too busy if we include too many controls there.

**Comment by swillisart** [comment]
> What do you think of the design proposed regarding the Outliner menu?
> What do you think of the design proposed regarding the Viewport menu?

This design works and is good to see everything available. I think It makes more intuitive sense from a Maya standpoint to make the context menus keyboard modifier aware. If there are not any existing plans for new context-menu’s **Shift+Right-Click** could be the way to avoid duplicate menu creation / nesting.

Maya’s existing marking menus and selections already work this way. Where different combinations of Ctrl, Shift and Ctrl+Shift each have their own set of context menu options.

**Shift+ Right-Click** could be operation on the multi-items selection and regular **Right-Click** could be exclusively on the picked / mouse over item.
I’m excited to see the progress here! And the presentation is perfect.

**Reply by womanyoyoyo-adsk** [reply]
Hi @swillisart , thanks for your feedback. I am going to rework it and post an update in the future. It has some issues. I'll keep you posted.

**Comment by womanyoyoyo-adsk** [comment]
FYI there are some issues with this design. I will work on it some more and post back. Thanks!

**Comment by womanyoyoyo-adsk** [comment]
Reworked this design. See new design here #3155
